src/metric/rank_metric.cu

```diff
@@ -0,0 +1,695 @@
+
+      dh::caching_device_vector<double> dbuf_pos(nitems, 0);
+      dh::caching_device_vector<double> dbuf_neg(nitems, 0);
+      auto *buf_pos_arr = dbuf_pos.data().get();
+      auto *buf_neg_arr = dbuf_neg.data().get();
+
+      dh::LaunchN(device_id, nitems, nullptr, [=] __device__(int idx) {
+        auto ctr = dlabels[dpreds_orig_pos[idx]];
+        // For ranking task, weights are per-group
+        // For binary classification task, weights are per-instance
+        const auto wt = dweights == nullptr ? 1.0f : dweights[dpreds_orig_pos[idx]];
+        buf_pos_arr[idx] = ctr * wt;
+        buf_neg_arr[idx] = (1.0f - ctr) * wt;
+        if (idx == nitems - 1 || dsorted_preds[idx] != dsorted_preds[idx + 1]) {
+          auto new_seg_idx = atomicAdd(seg_idx_ptr, 1);
+          ++new_seg_idx;  // As it is the old value; we want the last segment to have unique value
+          auto pred_val = dsorted_preds[idx];
+          do {
+            pred_seg_arr[idx] = new_seg_idx;
+            idx--;
+          } while (idx >= 0 && dsorted_preds[idx] == pred_val);
+        }
+      });
+
+      // Next, accumulate the positive and negative precisions for every prediction group
+      dh::caching_device_vector<double> sum_dbuf_pos(nitems, 0);
+      auto itr = thrust::reduce_by_key(thrust::cuda::par(alloc),
+                                       dpred_segs.begin(), dpred_segs.end(),  // Segmented by this
+                                       dbuf_pos.begin(),  // Individual precisions
+                                       thrust::make_discard_iterator(),  // Ignore unique segments
+                                       sum_dbuf_pos.begin());  // Write accumulated results here
+      auto nunique_preds = itr.second - sum_dbuf_pos.begin();
+      sum_dbuf_pos.resize(nunique_preds);
+
+      dh::caching_device_vector<double> sum_dbuf_neg(nunique_preds, 0);
+      itr = thrust::reduce_by_key(thrust::cuda::par(alloc),
+                                  dpred_segs.begin(), dpred_segs.end(),
+                                  dbuf_neg.begin(),
+                                  thrust::make_discard_iterator(),
+                                  sum_dbuf_neg.begin());
+      CHECK(itr.second - sum_dbuf_neg.begin() == nunique_preds);
+
+      // Find incremental sum for the positive precisions that is then used to
+      // compute incremental positive precision pair
+      dh::caching_device_vector<double> sum_npos(nunique_preds + 1, 0);
+      thrust::inclusive_scan(thrust::cuda::par(alloc),
+                             sum_dbuf_pos.begin(), sum_dbuf_pos.end(),
+                             sum_npos.begin() + 1);
+
+      dh::caching_device_vector<double> sum_nneg(nunique_preds, 0);
+      thrust::inclusive_scan(thrust::cuda::par(alloc),
+                             sum_dbuf_neg.begin(), sum_dbuf_neg.end(),
+                             sum_nneg.begin());
+
+      if (sum_npos.back() <= 0.0 || sum_npos.back() <= 0.0) {
```

Comment:
I think this use of back() implies 2 device to host copies, it would be more efficient to put this in a variable first then do the two checks.

Reply:
i think you caught a bug! one of them should be sum_nneg. i'll fix this after we revert #5365

Reply:
I would rather fix #5365 than revert.